File: Algorithm/c_or_cpp/charpter18_backtrace/level2/combination_sum.cpp
#include <iostream>
#include <vector>

using namespace std;

void dfs(vector<int>& c, int u, int target, vector<vector<int>>& res, vector<int>& path) {
    if (target < 0) {
        return;
    }
    if (target == 0) {
        res.push_back(path);
        return;
    }
    for (int i = u; i < c.size(); i++) {
        if (c[i] <= target) {
            path.push_back(c[i]);
            dfs(c, i, target - c[i], res, path);
            path.pop_back(); // 回溯，弹出当前节点
        }
    }
}

vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
    vector<vector<int>> res;
    vector<int> path;
    dfs(candidates, 0, target, res, path);
    return res;
}



int main() {
    vector<int> candidates = {10, 1, 2, 7, 6, 1, 5};
    int target = 8;
    vector<vector<int>> res = combinationSum(candidates, target);
    for (auto& path : res) {
        for (auto& num : path) {
            cout << num << " ";
        }
        cout << endl;
    }
    return 0;
}

File: Algorithm/Java/src/main/java/com/yugutou/charpter18_backtracking/level2/CombinationSum.java
package com.yugutou.charpter18_backtracking.level2;

import java.util.ArrayList;
import java.util.List;

/**
 * LeetCode77
 */
public class CombinationSum {

    List<List<Integer>> res = new ArrayList<>(); //记录答案
    List<Integer> path = new ArrayList<>();  //记录当前正在访问的路径

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        dfs(candidates, 0, target);
        return res;
    }

    public void dfs(int[] c, int u, int target) {
        if (target < 0) {
            return;
        }
        if (target == 0) {
            res.add(new ArrayList(path));
            return;
        }
        for (int i = u; i < c.length; i++) {
            if (c[i] <= target) {
                path.add(c[i]);
                //当前层将target减掉了一部分，也就是子结构只要找是否有满足(target -  c[i])就可以了
                dfs(c, i, target - c[i]); // 因为可以重复使用，所以还是i
                path.remove(path.size() - 1); //回溯
            }
        }
    }

    public static void main(String[] args) {
        int[] candidates = {2, 3, 6, 7};
        int target = 7;
        CombinationSum combinationSum = new CombinationSum();
        List<List<Integer>> result = combinationSum.combinationSum(candidates, target);
        System.out.println(result);
    }
}
